### PR TITLE
[SPARK-20456][Docs] Add examples for functions collection for pyspark

### PR DESCRIPTION
**Pull request by map222**
## What changes were proposed in this pull request?

This adds documentation to many functions in pyspark.sql.functions.py:
`upper`, `lower`, `reverse`, `unix_timestamp`, `from_unixtime`, `rand`, `randn`, `collect_list`, `collect_set`, `lit`
Add units to the trigonometry functions.
Renames columns in datetime examples to be more informative.
Adds links between some functions.

## How was this patch tested?

`./dev/lint-python`
`python python/pyspark/sql/functions.py`
`./python/run-tests.py --module pyspark-sql`

### CODE REVIEW

**Comment by map222** [comment]
@HyukjinKwon I ended up not making examples for the aggregate functions, as I didn't make a good dataframe to demonstrate them. I could add more examples for the string functions if you think that is a good idea. There are dozens of functions that could be documented, I'm not sure how far we want to go, or which ones need it.

**Comment by gatorsmile** [comment]
ok to test

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -153,7 +173,7 @@ def _():
 # math functions that take two arguments as input
 _binary_mathfunctions = {
     'atan2': 'Returns the angle theta from the conversion of rectangular coordinates (x, y) to' +
-             'polar coordinates (r, theta).',
+             'polar coordinates (r, theta). Units in radians.',
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I am unclear we should note this for every instance and users really gets confused as I see these use Scala/Java's built-in library. I wonder if there is an example that supports this differently in other libraries?

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -206,17 +226,20 @@ def _():
 @since(1.3)
 def approxCountDistinct(col, rsd=None):
     """
-    .. note:: Deprecated in 2.1, use approx_count_distinct instead.
+    .. note:: Deprecated in 2.1, use :func:`approx_count_distinct instead`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Probably `` :func:`approx_count_distinct` ``?

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -397,7 +420,7 @@ def input_file_name():
 
 @since(1.6)
 def isnan(col):
-    """An expression that returns true iff the column is NaN.
+    """An expression that returns true if the column is NaN.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I think "iff" is the abbreviation for "if and only if". I don't think it is worth changing.

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -456,7 +479,7 @@ def monotonically_increasing_id():
 def nanvl(col1, col2):
     """Returns col1 if it is not NaN, or col2 if col1 is NaN.
 
-    Both inputs should be floating point columns (DoubleType or FloatType).
+    Both inputs should be floating point columns (:class:`DoubleType` or FloatType).
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I think we should link both `DoubleType` and `FloatType ` all or not.

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -1120,12 +1159,12 @@ def from_utc_timestamp(timestamp, tz):
 @since(1.5)
 def to_utc_timestamp(timestamp, tz):
     """
-    Given a timestamp, which corresponds to a certain time of day in the given timezone, returns
-    another timestamp that corresponds to the same time of day in UTC.
+    Given a `timestamp`, which corresponds to a time of day in the timezone `tz`,
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Should this be ``` ``timestamp`` ``` not `` `timestamp` ``?

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -67,9 +67,16 @@ def _():
     _.__doc__ = 'Window function: ' + doc
     return _
 
+_lit_doc = """
+    Creates a :class:`Column` of literal value. Supports basic types like :class:`IntegerType`,
+    :class:`FloatType`, :class:`BooleanType`, and :class:`StringType`
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I would like to keep this identical with the one in `functions.scala` to reduce overhead when someone sweeps the same documentation changes across APIs in other languages.

If the additional informations are Python-specific, let's add it in `::note`.

File: python/pyspark/sql/functions.py [hunk 7]
```diff
@@ -793,8 +824,8 @@ def date_format(date, format):
     .. note:: Use when ever possible specialized functions like `year`. These benefit from a
         specialized implementation.
 
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['a'])
-    >>> df.select(date_format('a', 'MM/dd/yyy').alias('date')).collect()
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['time'])
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Okay. I guess it is a documentation improvement to use a bit more meaningful name over an arbitrary name `a`. Let's match these to existing names such as `ts` or `t` (abbreviation for timestamp) or `dt` (abbreviation for `datetime.datetime`).

File: python/pyspark/sql/functions.py [hunk 8]
```diff
@@ -910,8 +941,8 @@ def weekofyear(col):
     """
     Extract the week number of a given date as integer.
 
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['a'])
-    >>> df.select(weekofyear(df.a).alias('week')).collect()
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['time'])
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Let's use `d` for `DateType` or `datetime.date` similarly with other existing names.

**Review by HyukjinKwon — COMMENTED** [review]
I have not finished my review. I will take a look again.
Let's fix up lint check failures.

In general, I think I am not sure of changing the names to `time`. It does not look particularly better to me but unnecessary changes.

Also, let's check out the contents in Scala/R ones. I think we should keep them identical in most cases.

If there are Python-specific contents, let's use `::note`. If it is R-specific, ` Note: `. If it is Scala/Java specific, `@note`.

**Comment by HyukjinKwon** [comment]
(Thank you @gatorsmile for triggering the test)

File: python/pyspark/sql/functions.py [hunk 9]
```diff
@@ -409,7 +432,7 @@ def isnan(col):
 
 @since(1.6)
 def isnull(col):
-    """An expression that returns true iff the column is null.
+    """An expression that returns true if the column is null.
```
Comment (gatorsmile) [inline python/pyspark/sql/functions.py]: `the column`? This is misleading. We should make the Python documents consistent with what we did in Scala. 
For example, `isNull` in Scala APIs is described as
> Returns true if `expr` is null, or false otherwise.

Ref: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala#L280

**Comment by gatorsmile** [comment]
Could you please check the documents we did in Scala APIs? It sounds like we forgot to update the Python function descriptions when we did the change in the Scala APIs.

File: python/pyspark/sql/functions.py (hunk 7 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: I am replacing all of the timestamps of form `yyyy-MM-dd` with `dt`, and all of the `yyyy-MM-dd HH:mm:ss` with ts, with some exceptions. Hopefully this addresses this point.

File: python/pyspark/sql/functions.py (hunk 5 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: I had a quick question about this. The documentation seems pretty inconsistent about when to use \` vs \`\`. For example, in [approx_count_distinct](http://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.functions.approx_count_distinct), the parameter `col` is referred to with ``, while in [rpad](http://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.functions.rpad), the parameters are referred to with `.

Numpy appears to use single backticks for parameters:
https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections

Is there a reference for this?

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: Most libraries seem to default to radians. However, I checked the R, numpy, and MATLAB notes for common trigonometry functions, and they all note the units in the function documentation, e.g.:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.sin.html
https://www.mathworks.com/help/matlab/ref/sin.html

**Comment by map222** [comment]
@gatorsmile I checked four functions, `approx_count_distinct`, `coalesce`, `covar_samp`, and `countDistinct`, comparing the python and Scala documentation. None of them are the same. My guess is that the python docs differ for most functions.

**Comment by gatorsmile** [comment]
@map222 Unfortunately, our PySpark did not follow what we did in Scala. Will review it more carefully in the future. Thanks!

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I see. What do you think about adding this in `:param`?

File: python/pyspark/sql/functions.py (hunk 5 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: No, I think we have a rule about this up to my knowledge. Thank you for the pointers and looking into this. Let's follow the majority then for now.

Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: No, I don't think we have a rule about this up to my knowledge. Thank you for the pointers and looking into this. Let's follow the majority then for now.

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: I added `:param:` to both `acos` and `asin` using two methods. What do you think?

**Comment by map222** [comment]
Most recent commit reverts the "if" to "iff", changes all the backticks for column names to single backtick, and tried new `:param:` option for angle columns.

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: @HyukjinKwon Any opinion on how to do the `:param:` for the units?

**Comment by map222** [comment]
Could I get a review for this? I think the only remaining questions is whether (and how) to note the units for the trigonometry functions, like here:
https://github.com/map222/spark/blob/dd7a3971275aae97f9adfd82166df23b038eb950/python/pyspark/sql/functions.py#L100-L105

**Comment by map222** [comment]
@HyukjinKwon @gatorsmile bump

**Comment by gatorsmile** [comment]
Also cc @ueshin

**Comment by HyukjinKwon** [comment]
Will take a look to help too soon. Sorry for the delay.

**Review by ueshin — COMMENTED** [review]
I added some minor style-related comments.

File: python/pyspark/sql/functions.py [hunk 10]
```diff
@@ -189,15 +210,15 @@ def _():
 }
 
 for _name, _doc in _functions.items():
-    globals()[_name] = since(1.3)(_create_function(_name, _doc))
+    globals()[_name] = since(1.3)(ignore_unicode_prefix(_create_function(_name, _doc)))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Hm, this then applies `ignore_unicode_prefix` to all docs ... This looks only added for `_lit_doc`. Could we simply avoid the unicode example in the lit doc rather than applying this to all function docs?

File: python/pyspark/sql/functions.py [hunk 11]
```diff
@@ -189,15 +210,15 @@ def _():
...
 for _name, _doc in _functions_1_4.items():
     globals()[_name] = since(1.4)(_create_function(_name, _doc))
 for _name, _doc in _binary_mathfunctions.items():
     globals()[_name] = since(1.4)(_create_binary_mathfunction(_name, _doc))
 for _name, _doc in _window_functions.items():
     globals()[_name] = since(1.6)(_create_window_function(_name, _doc))
 for _name, _doc in _functions_1_6.items():
-    globals()[_name] = since(1.6)(_create_function(_name, _doc))
+    globals()[_name] = since(1.6)(ignore_unicode_prefix(_create_function(_name, _doc)))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: This one looks also the same instance. Let's avoid adding unicode examples in `_collect_list_doc` and `_collect_set_doc` for now if we are all fine.

File: python/pyspark/sql/functions.py [hunk 12]
```diff
@@ -206,17 +227,20 @@ def _():
...
-    .. note:: Deprecated in 2.1, use approx_count_distinct instead.
+    .. note:: Deprecated in 2.1, use :func:`approx_count_distinct` instead.
     """
     return approx_count_distinct(col, rsd)
 
 
 @since(2.1)
 def approx_count_distinct(col, rsd=None):
-    """Returns a new :class:`Column` for approximate distinct count of ``col``.
+    """Returns a new :class:`Column` for approximate distinct count of column `col`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Sounds like it also needs leading `Aggregate function: ` doc - https://github.com/apache/spark/blob/3a840048ed3501e06260b7c5df18cc0bbdb1505c/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L225.

File: python/pyspark/sql/functions.py [hunk 13]
```diff
@@ -206,17 +227,20 @@ def _():
...
 
-    >>> df.agg(approx_count_distinct(df.age).alias('c')).collect()
-    [Row(c=2)]
+    :param rsd: Residual (float). The approximate count will be within this fraction of the true
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: For this one, let's add the contents in Scala/R one too - https://github.com/apache/spark/blob/3a840048ed3501e06260b7c5df18cc0bbdb1505c/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L245 and https://github.com/apache/spark/blob/7f203a248f94df6183a4bc4642a3d873171fef29/R/pkg/R/functions.R#L2243

This one in Scala/R look already having the param contents to match with.

File: python/pyspark/sql/functions.py [hunk 14]
```diff
@@ -793,9 +824,9 @@ def date_format(date, format):
     .. note:: Use when ever possible specialized functions like `year`. These benefit from a
         specialized implementation.
 
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['a'])
-    >>> df.select(date_format('a', 'MM/dd/yyy').alias('date')).collect()
-    [Row(date=u'04/08/2015')]
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['dt'])
+    >>> df.select(date_format('dt', 'MM/dd/yyy').alias('dt2')).collect()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I think `date` was also fine.

File: python/pyspark/sql/functions.py [hunk 15]
```diff
@@ -962,9 +993,9 @@ def add_months(start, months):
     """
     Returns the date that is `months` months after `start`
 
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['d'])
-    >>> df.select(add_months(df.d, 1).alias('d')).collect()
-    [Row(d=datetime.date(2015, 5, 8))]
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['dt'])
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I think changing arbitrary name to `dt` or `ts` is fine but not worth changing `t`, `timestamp`, `d` or `date`. These are already readable.

File: python/pyspark/sql/functions.py [hunk 16]
```diff
@@ -1254,23 +1294,41 @@ def hash(*cols):
...
+    'reverse': _reverse_doc,
     'ltrim': 'Trim the spaces from left end for the specified string value.',
     'rtrim': 'Trim the spaces from right end for the specified string value.',
     'trim': 'Trim the spaces from both ends for the specified string column.',
 }
 
 
 for _name, _doc in _string_functions.items():
-    globals()[_name] = since(1.5)(_create_function(_name, _doc))
+    globals()[_name] = since(1.5)(ignore_unicode_prefix(_create_function(_name, _doc)))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Hm.. can we avoid applying this to all docs too? One way is just to take out the examples.  These string function examples looks easy and straightforward.

**Review by HyukjinKwon — COMMENTED** [review]
I also left some comments.

File: python/pyspark/sql/functions.py (hunk 10 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: `ignore_unicode_prefix` is necessary to get the tests to pass both python 2 and python 3. I had problems without it on my previous documentation PR: https://github.com/apache/spark/pull/17469#issuecomment-295896057

File: python/pyspark/sql/functions.py [hunk 17]
```diff
@@ -92,14 +98,16 @@ def _():
...
     'acos': 'Computes the cosine inverse of the given value; the returned angle is in the range' +
-            '0.0 through pi.',
-    'asin': 'Computes the sine inverse of the given value; the returned angle is in the range' +
-            '-pi/2 through pi/2.',
-    'atan': 'Computes the tangent inverse of the given value.',
+            '0.0 through pi.\n\n:param col: float column, units in radians.',
+    'asin': """Computes the sine inverse of the given value; the returned angle is in the range
+            -pi/2 through pi/2.
+
+            :param col: float column, units in radians.""",
```
Comment (map222) [inline python/pyspark/sql/functions.py]: I was trying two different ways to see which worked best. I now use the `:param col:` version

File: python/pyspark/sql/functions.py (hunk 10 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Yea, I didn't mean to not use but avoid applying all other docs (more specifically the ones in `_functions`).  This basically replaces u prefix via regex and rather a workaround IMHO. So, I think we should avoid applying this to other doctests not using unicode.

Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: And .. the easiest way I think is to take out the examples or not use unicode in the doctests for now as `ignore_unicode_prefix` requires a function.

File: python/pyspark/sql/functions.py (hunk 15 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: In the original example, the column name `d` is reused for the new column. I think using a new column name helps differentiate the original date from the new date.

File: python/pyspark/sql/functions.py (hunk 16 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: I think it is helpful to be clear on how `lower` and `upper` should be used. This week I tried to use `lower` as a column function, rather than as an outside function. I can remove the docstrings if you think avoiding `ignore_unicode_prefix` is important.

Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: In that way, we should add all examples for each function here. I am not too sure if we need them all for now. At least per - https://issues.apache.org/jira/browse/SPARK-17963?focusedCommentId=15581022&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15581022.

File: python/pyspark/sql/functions.py (hunk 15 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Okay, but let's sweep the other instances here if there are.

**Comment by holdenk** [comment]
Thanks for working on this :) More complete documentation is certainly useful. Looking at the "How was this patch tested?" section you might also want to build the docs (there are instructions in `docs/README.md`) to see how stuff is formatted.

**Comment by map222** [comment]
Ok, I removed the `ignore_unicode_prefix`, and it didn't have any test problems. Also removed the date column renames where the original name was ok. Finally, removed the string function documentation.

**Comment by map222** [comment]
@HyukjinKwon I could not get it to pass the tests without the unicode and `ignore_unicode_prefix`

File: python/pyspark/sql/functions.py [hunk 18]
```diff
@@ -147,7 +173,7 @@ def _():
 # math functions that take two arguments as input
 _binary_mathfunctions = {
     'atan2': 'Returns the angle theta from the conversion of rectangular coordinates (x, y) to' +
-             'polar coordinates (r, theta).',
+             'polar coordinates (r, theta). Units in radians.',
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I think we should ...

1. copy and paste `Units in radians.` to both https://github.com/apache/spark/blob/e5387018e76a9af1318e78c4133ee68232e6a159/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L1342 and https://github.com/apache/spark/blob/8965fe764a4218d944938aa4828072f1ad9dbda7/R/pkg/R/functions.R#L2030 just to be consistent

2. Use Python specific notation to describe this here for now.

3. Just revert this change for now.

File: python/pyspark/sql/functions.py [hunk 19]
```diff
@@ -1258,7 +1303,7 @@ def hash(*cols):
                'uppercase. Words are delimited by whitespace.',
     'lower': 'Converts a string column to lower case.',
     'upper': 'Converts a string column to upper case.',
-    'reverse': 'Reverses the string column and returns it as a new string column.',
+    'reverse': 'Reverses a string column and returns it as a new string column.',
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: ditto. I guess we should match this change to R/Scala ones too.

File: python/pyspark/sql/functions.py [hunk 20]
```diff
@@ -109,15 +118,33 @@ def _():
...
+    'toDegrees': '.. note:: Deprecated in 2.1, use :func:`degrees` instead.',
+    'toRadians': '.. note:: Deprecated in 2.1, use :func:`radians` instead.',
     'bitwiseNOT': 'Computes bitwise not.',
 }
 
+_collect_list_doc = """
+    Aggregate function: returns a list of objects with duplicates.
+
+    >>> df2 = spark.createDataFrame([('Alice', 2), ('Bob', 5), ('Alice', 99)], ('name', 'age'))
+    >>> df2.agg(collect_list('name')).collect()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I think we can avoid `ignore_unicode_prefix` if we call this function with `age` for this and the one below.

File: python/pyspark/sql/functions.py [hunk 21]
```diff
@@ -67,9 +67,15 @@ def _():
     _.__doc__ = 'Window function: ' + doc
     return _
 
+_lit_doc = """
+    Creates a :class:`Column` of literal value.
 
+    >>> df.withColumn('height', lit(5)).withColumn('spark_user', lit(True)).collect()
+    [Row(age=2, name=u'Alice', height=5, spark_user=True),
+     Row(age=5, name=u'Bob', height=5, spark_user=True)]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Here too, I think we can just do something like `df.select(hight, lit(5)).collect()` or `show()` to avoid `ignore_unicode_prefix`.

File: python/pyspark/sql/functions.py [hunk 22]
```diff
@@ -200,17 +226,20 @@ def _():
...
-    .. note:: Deprecated in 2.1, use approx_count_distinct instead.
+    .. note:: Deprecated in 2.1, use :func:`approx_count_distinct` instead.
     """
     return approx_count_distinct(col, rsd)
 
 
 @since(2.1)
-def approx_count_distinct(col, rsd=None):
-    """Returns a new :class:`Column` for approximate distinct count of ``col``.
+def approx_count_distinct(col, rsd=0.05):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Why do we change this to `0.05`? We then should change the default values in both places in `HyperLogLogPlusPlus` and  here if we happen to change this in any event. Let's leave it `None`.

File: python/pyspark/sql/functions.py [hunk 23]
```diff
@@ -267,8 +296,7 @@ def coalesce(*cols):
 
 @since(1.6)
 def corr(col1, col2):
-    """Returns a new :class:`Column` for the Pearson Correlation Coefficient for ``col1``
-    and ``col2``.
+    """Returns a new :class:`Column` for the Pearson Correlation Coefficient for `col1` and `col2`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Honestly, this is rather unnecessary change. Both before and after are fine and do not affect the documentation rendering.

File: python/pyspark/sql/functions.py (hunk 14 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: `date` to `dt2` sounds unnecessary change.

File: python/pyspark/sql/functions.py [hunk 24]
```diff
@@ -969,8 +1005,8 @@ def months_between(date1, date2):
     """
     Returns the number of months between date1 and date2.
 
-    >>> df = spark.createDataFrame([('1997-02-28 10:30:00', '1996-10-30')], ['t', 'd'])
-    >>> df.select(months_between(df.t, df.d).alias('months')).collect()
+    >>> df = spark.createDataFrame([('1997-02-28 10:30:00', '1996-10-30')], ['date1', 'date2'])
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: `t` and `d` looks already fine.

File: python/pyspark/sql/functions.py [hunk 25]
```diff
@@ -1073,12 +1109,17 @@ def last_day(date):
...
+@ignore_unicode_prefix
 @since(1.5)
 def from_unixtime(timestamp, format="yyyy-MM-dd HH:mm:ss"):
     """
     Converts the number of seconds from unix epoch (1970-01-01 00:00:00 UTC) to a string
     representing the timestamp of that moment in the current system time zone in the given
     format.
+
+    >>> time_df = spark.createDataFrame([(1428476400, )], ['unix_time'])
+    >>> time_df.select(from_unixtime('unix_time').alias('ts') ).collect()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: nit: `) )` -> `))`.

File: python/pyspark/sql/functions.py [hunk 26]
```diff
@@ -1105,9 +1150,9 @@ def from_utc_timestamp(timestamp, tz):
     Given a timestamp, which corresponds to a certain time of day in UTC, returns another timestamp
     that corresponds to the same time of day in the given timezone.
 
-    >>> df = spark.createDataFrame([('1997-02-28 10:30:00',)], ['t'])
-    >>> df.select(from_utc_timestamp(df.t, "PST").alias('t')).collect()
-    [Row(t=datetime.datetime(1997, 2, 28, 2, 30))]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: `t` to `ts` looks unnecessary change here and the one below.

File: python/pyspark/sql/functions.py [hunk 27]
```diff
@@ -312,7 +338,7 @@ def covar_samp(col1, col2):
 
 @since(1.3)
 def countDistinct(col, *cols):
-    """Returns a new :class:`Column` for distinct count of ``col`` or ``cols``.
+    """Returns a new :class:`Column` for distinct count of `col` or `cols`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I guess we did not decide which one of `` `...` `` or ``` ``...`` ``` is more preferred and right. Let's avoid sweeping them. Adding new one is fine (and I usually follow majority in this case) but changing the existing one might be not if we don't know which one is not preferred. We might happen to change it forward and backward in time.

**Review by HyukjinKwon — COMMENTED** [review]
I believe these are all from me.

File: python/pyspark/sql/functions.py (hunk 22 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Okay, I double checked some APIs have them but let's leave this out. It sounds requiring fixing/checking logic below and the deprecated one above. I assume this should be doc only change PR.

**Comment by map222** [comment]
@HyukjinKwon I think I addressed all of your comments. Thank you for your detailed review!

File: python/pyspark/sql/functions.py (hunk 24 above)
Comment (map222) [inline python/pyspark/sql/functions.py]: I reverted all the other ones, but I like this one because `date1` and `date2` match the docstring. Is that reasonable?

File: R/pkg/R/functions.R [hunk 28]
```diff
@@ -426,7 +426,7 @@ setMethod("covar_pop", signature(col1 = "characterOrColumn", col2 = "characterOr
 
 #' cos
 #'
-#' Computes the cosine of the given value.
+#' Computes the cosine of the given value. Units in radians.
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: looks like this will conflict with #18371

Comment (HyukjinKwon) [inline R/pkg/R/functions.R]: Yes, if that one is merged first, we should resolve the conflict here. If this one gets merged first, that one should resolve the conflict. Describing the same information across APIs looks reasonable to me.

Comment (HyukjinKwon) [inline R/pkg/R/functions.R]: BTW, just while you are here @felixcheung, does this change itself look okay to you?

File: python/pyspark/sql/functions.py [hunk 29]
```diff
@@ -200,17 +225,20 @@ def _():
...
-    .. note:: Deprecated in 2.1, use approx_count_distinct instead.
+    .. note:: Deprecated in 2.1, use :func:`approx_count_distinct` instead.
     """
     return approx_count_distinct(col, rsd)
 
 
 @since(2.1)
 def approx_count_distinct(col, rsd=None):
-    """Returns a new :class:`Column` for approximate distinct count of ``col``.
+    """Aggregate function. Returns a new :class:`Column` for approximate distinct count of column `col`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: little nit `. R` -> `: r`.

File: python/pyspark/sql/functions.py [hunk 30]
```diff
@@ -282,8 +309,7 @@ def corr(col1, col2):
 
 @since(2.0)
 def covar_pop(col1, col2):
-    """Returns a new :class:`Column` for the population covariance of ``col1``
-    and ``col2``.
+    """Returns a new :class:`Column` for the population covariance of `col1` and `col2`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: little nit: `` `col1` `` -> ``` ``col1`` ``` and the same one below.

File: python/pyspark/sql/functions.py [hunk 31]
```diff
@@ -1116,12 +1160,12 @@ def from_utc_timestamp(timestamp, tz):
 @since(1.5)
 def to_utc_timestamp(timestamp, tz):
     """
-    Given a timestamp, which corresponds to a certain time of day in the given timezone, returns
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Let's revert this change or do this the same thing to Scala/R too if it matters.

File: python/pyspark/sql/functions.py [hunk 32]
```diff
@@ -1258,7 +1302,7 @@ def hash(*cols):
                'uppercase. Words are delimited by whitespace.',
     'lower': 'Converts a string column to lower case.',
     'upper': 'Converts a string column to upper case.',
-    'reverse': 'Reverses the string column and returns it as a new string column.',
+    'reverse': 'Reverses a string column and returns it as a new string column.',
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: ditto. I guess we should revert this change or match this change to R/Scala ones too if it matters.

File: python/pyspark/sql/functions.py [hunk 33]
```diff
@@ -473,10 +503,15 @@ def rand(seed=None):
     return Column(jc)
 
 
+@ignore_unicode_prefix
 @since(1.4)
 def randn(seed=None):
     """Generates a column with independent and identically distributed (i.i.d.) samples from
     the standard normal distribution.
+
+    >>> df.withColumn('randn', randn(seed=42) ).collect()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: little nit: `) )` -> `))`

File: python/pyspark/sql/functions.py [hunk 34]
```diff
@@ -95,10 +100,13 @@ def _():
             '0.0 through pi.',
     'asin': 'Computes the sine inverse of the given value; the returned angle is in the range' +
             '-pi/2 through pi/2.',
-    'atan': 'Computes the tangent inverse of the given value.',
+    'atan': 'Computes the tangent inverse of the given value; the returned angle is in the range' +
+            '-pi/2 through pi/2',
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Let's revert this change or add the same information in Scala/R too.

**Review by HyukjinKwon — COMMENTED** [review]
It looks good to me except for few minor comments and the comment above ^.

File: python/pyspark/sql/functions.py [hunk 35]
```diff
@@ -1073,12 +1108,17 @@ def last_day(date):
...
 
+@ignore_unicode_prefix
 @since(1.5)
 def from_unixtime(timestamp, format="yyyy-MM-dd HH:mm:ss"):
     """
     Converts the number of seconds from unix epoch (1970-01-01 00:00:00 UTC) to a string
     representing the timestamp of that moment in the current system time zone in the given
     format.
+
+    >>> time_df = spark.createDataFrame([(1428476400, )], ['unix_time'])
```
Comment (ueshin) [inline python/pyspark/sql/functions.py]: little nit: `(1428476400, )` -> `(1428476400,)`

**Review by ueshin — COMMENTED** [review]
LGTM except for a minor comment and comments above.

**Comment by map222** [comment]
I think I addressed everything.

**Comment by ueshin** [comment]
Looks like there is a merge conflict. Can you fix it?

**Comment by map222** [comment]
Do I just have to `git rebase` on master?

**Comment by ueshin** [comment]
Yes, or you can merge master and fix the conflict.

File: R/pkg/R/functions.R [hunk 36]
```diff
@@ -148,7 +148,8 @@ setMethod("asin",
 
 #' atan
 #'
-#' Computes the tangent inverse of the given value.
+#' Computes the tangent inverse of the given value; the returned angle is in the range
+#' -pi/2 through pi/2
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: this has been moved in master - you should see when you rebase

File: R/pkg/R/functions.R [hunk 37]
```diff
@@ -1900,8 +1901,8 @@ setMethod("year",
 
 #' @details
 #' \code{atan2}: Returns the angle theta from the conversion of rectangular coordinates
-#' (x, y) to polar coordinates (r, theta).
-#'
+#' (x, y) to polar coordinates (r, theta). Units in radians.
+#
```
Comment (HyukjinKwon) [inline R/pkg/R/functions.R]: Sorry... missing `'`...

**Review by HyukjinKwon — APPROVED** [review]
It looks https://github.com/apache/spark/pull/17865/files#r123964695 is missed .. LGTM too.

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Review by felixcheung — COMMENTED** [review]
LGTM. 
how about https://github.com/apache/spark/pull/17865/files#r114929075?

**Comment by map222** [comment]
I switched the examples to numeric examples so that we could avoid using `ignore_unicode_prefix`, as requested by @HyukjinKwon in [this commit](https://github.com/apache/spark/pull/17865/commits/0ab91297c09628377fffc04b3f54216bf2581a89#diff-f5295f69bfbdbf6e161aed54057ea36dL144)

**Comment by ueshin** [comment]
How about https://github.com/apache/spark/pull/17865/files#r123964695 ?

**Comment by HyukjinKwon** [comment]
Yea. For https://github.com/apache/spark/pull/17865#issuecomment-313303801, just to explain the context (as it is a bit confusing even to me), it was suggested to use `ignore_unicode_prefix` for single instance when the example has an unicode in the doctest _or_ to remove the unicode in the doctest if possible (either should be fine).

I also suggested to not use it for unrelated docstrings - https://github.com/apache/spark/pull/17865#discussion_r121539034.

**Review by felixcheung — COMMENTED** [review]
any more thought on this?

**Comment by ueshin** [comment]
LGTM.

**Comment by HyukjinKwon** [comment]
LGTM too.

**Comment by felixcheung** [comment]
merged to master. thanks!
